File: LeetCode_Algorithm/127(Word_Ladder)/127.js
var ladderLength = function(beginWord, endWord, wordList) {
  // sol1 2164ms 11.6%
  // let graph = new Map()
  // for (let word of wordList) {
  //   for (let idx in word) {
  //     let temp = [...word]; temp[idx] = '_';
  //     let key = temp.join('')
  //     if (graph[key] == undefined) {
  //       graph[key] = [word]
  //     } else {
  //       graph[key].push(word)
  //     }
  //   }
  // }
  // let queue = [[beginWord, 0]]; let visited = []
  // while (queue.length > 0) {
  //   let temp = queue.shift()
  //   let word = temp[0]; let cnt = temp[1]
  //   if (word == endWord) {
  //     return cnt+1
  //   }
  //   if (visited.indexOf(word) > -1) {
  //     continue
  //   }
  //   visited.push(word)
  //   for (let idx in word) {
  //     let lst_word = [...word]; lst_word[idx] = '_';
  //     let key_word = lst_word.join('')
  //     if (graph[key_word] != undefined) {
  //       for (let temp_word of graph[key_word]) {
  //         queue.push([temp_word, cnt+1])
  //       }
  //     }
  //   }
  // }
  // return 0

  // sol2 169ms 86% 63.5MB 9%
  let graph = new Map()
  for (let word of wordList) {
    for (let i = 0; i < word.length; i++) {
      let key = word.slice(0, i) + '*' + word.slice(i+1, )
      if (graph[key] == undefined)
        graph[key] = []
      graph[key].push(word)
    }
  }
  let queue = [[beginWord, 1]]; let seen = new Set([beginWord])
  while (queue.length) {
    let [word, cnt] = queue.shift()
    if (word == endWord)
      return cnt
    for (let i = 0; i < word.length; i++) {
      let key = word.slice(0, i) + '*' + word.slice(i+1, )
    if (graph[key] != undefined) {
        for (let w of graph[key]) {
            if (!seen.has(w)) {
              seen.add(w)
              queue.push([w, cnt+1])
            }
        }
      }
    }
 }
 return 0
};


console.log(ladderLength("hit", "dog", ["hot","dog","dot"]))

File: LeetCode_Algorithm/127(Word_Ladder)/127.py
from typing import List
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:

        if not beginWord or not endWord or not wordList:
            return 0
        if endWord not in wordList:
            return 0
        # sol1 4941 5%
        # graph를 그려서 가능한 모든 값에 key를 매핑하고 word를 매핑한다.
        # graph = {};
        # for word in wordList:
        #     for idx in range(len(word)):
        #         temp = list(word); temp[idx] = "_"
        #         key = ''.join(temp)
        #         if key not in graph:
        #             graph[key] = [word]
        #         else:
        #             graph[key].append(word)
        # # bfs로 찾아가면서 목표에 도달하면 cnt+1 아니면 계속 루프
        # from collections import deque
        # queue = deque([(beginWord, 0)]); visited = []
        # while queue:
        #     word, cnt = queue.popleft()
        #     if word == endWord:
        #         return cnt+1
        #     if word in visited:
        #         continue
        #     visited.append(word)
        #     for idx in range(len(word)):
        #         temp = list(word); temp[idx] = "_"
        #         key = ''.join(temp)
        #         if graph.get(key):
        #             for word_key in graph[key]:
        #                 queue.append((word_key, cnt+1))
        # return 0

        # sol2 126ms 91% 18.4MB 19%
        from collections import defaultdict, deque
        graph = defaultdict(list)
        for word in wordList:
            for i in range(len(word)):
                key = word[:i] + '*' + word[i+1:]
                graph[key].append(word)
        queue = deque([(beginWord, 1)]); seen = set([beginWord])
        while queue:
            word, cnt = queue.popleft()

            if word == endWord:
                return cnt
            for i in range(len(word)):
                key = word[:i] + '*' + word[i+1:]
                for w in graph[key]:
                    if w not in seen:
                        seen.add(w)
                        queue.append((w, cnt+1))
                graph[key] = []
        return 0

        
print(Solution().ladderLength("a", "c", ["a", "b", "c"]))